astropy/table/table.py

```diff
@@ -905,10 +908,7 @@ def _convert_data_to_col(self, data, copy=True, default_name=None, dtype=None, n
             # gets upgraded to MaskedColumn, but the converse (pre-4.0) behavior
             # of downgrading from MaskedColumn to Column (for non-masked table)
             # does not happen.
-            if issubclass(self.ColumnClass, data.__class__):
-                col_cls = self.ColumnClass
-            else:
-                col_cls = data.__class__
+            col_cls = self._get_col_cls_for_table(data)
```

Comment:
This refactor (here and below) and change in behavior was necessary to get the hstack conversion working.  But in general it is a good thing for more consistent behavior that works with the advertised table API of allowing custom default `Column` and `MaskedColumn` classes.